R/chat-structured.R

```diff
@@ -84,3 +68,33 @@ convert_from_type <- function(x, type) {
     x
   }
 }
+
+list_to_atomic <- function(x, type) {
+  r_type <- switch(
+    type,
+    boolean = "logical",
+    integer = "integer",
+    number = "double",
+    string = "character",
+    cli::cli_abort("Unknown type {type}", .internal = TRUE)
+  )
+
+  if (length(x) == 0) {
+    return(vector(r_type, 0))
+  }
+
+  types <- map_chr(x, typeof)
+  if (r_type == "integer") {
+    x[types == "double"] <- lapply(x[types == "double"], as.integer)
+    types[types == "double"] <- "integer"
+  } else if (r_type == "double") {
+    x[types == "integer"] <- lapply(x[types == "integer"], as.double)
+    types[types == "integer"] <- "double"
+  }
+
+  # also captures NULLs
+  wrong_type <- types != r_type
+  x[wrong_type] <- list(`mode<-`(NA, r_type))
```

Comment:
How strict do you want to be here? No warning at all for unexpected types? That feels kind of scary. I would have thought you'd handle `NULL`s explicitly up front, and then do a final check for invalid types at the end and error if any are still invalid.

Reply:
Type is already checked at the beginning?

Reply:
I think I'm talking about what happens if your `r_type` is `"integer"` and in `x` you have a string

Reply:
Oh yeah, I do.

Reply:
I did that and made it more clear that's the goal with a comment

Reply:
Seems reasonable!